### ISSUE TITLE
get-cabal-configuration returning wrong version of stack build directory

### ISSUE THREAD

**Issue by maxigit** [flycheck/flycheck-haskell#57]
I've updated cabal-install to 1.22.8 using `cabal install cabal` (or equivalent, I don't remember exactly the command) and this break flycheck : I get a `can't find cabal_macros.h`. The reason is `get-cabal-configuration.hs` generate the path to find this file in `.stack-work/dist/x86_64-linux/Cabal-1.22.8.0/build/autogen` but this directory doesn't exist but is `.stack-work/dist/x86_64-linux/Cabal-1.22.5./build/autogen` instead. Basically stack doesn't know I've upgraded cabal to 1.22.8.   and still use 1.22.5.

The workaround is to upgrade the stack setup with `stack setup --upgrade-cabal` .
The correct way to find the stack dist directory seems to be using `stack path --dist-dir`.

**Comment by lunaryorn** [comment]
@maxigit Thanks, but I'm not sure what the issue is here?  What should Flycheck do different?

**Comment by maxigit** [comment]
Set the stack dist path properly . The current way to find the path using the get-cabal-configuration is wrong. We should use the path given by  the command `stack path --project-root` instead.

**Comment by lunaryorn** [comment]
@maxigit I must admit that I'm not convinced that Flycheck is at fault.  

We're not guessing the Cabal version in `get-cabal-configuration.hs` after all; the Cabal version comes from the cabal package that we load, and provided that `get-cabal-configuration.hs` runs through `stack` (which is the default) we should the Cabal version Stack has for the current project and not the global one.  I may be mistaken, though, it's a long time that I looked at this.

Anyway, I do not have the time to debug and address this issue myself.  It'd be great if you could take a look and open a pull request.

**Comment by maxigit** [comment]
The point is that the the cabal version run through stack is incorrect. What matters is the cabal version corresponding to ghc version used by stack. I don't really have the time to fix it myself. Even though the fix is in theory quite simple (use `stack path --dist-dir` instead of trying to guess it) it seems that it needs major refactoring. 
Moreover, the fix is quite simple (see original post) which is the main reason I raised this issue : so that people having this issue can fix it.

**Comment by lunaryorn** [comment]
Our Cabal version comes from `stack runghc`.  How can that _not_ be the Cabal version that's associated with the GHC being used for the current project?

**Comment by maxigit** [comment]
Well, I'm not sure what Cabal version `stack runghc` version is picking but it's definitely not the correct one.
I expect it to be the version from the cabal corresponding to the stackage snapshot, but it's even not that.

`get-cabal-configuration.hs` uses `Distribution.Simple.Utils(cabalVersion)`. If try to import this in ghci (stack ghci) I get an error telling me it belongs to an hidden package.
stack doesn't use this Distribution.Simple.Utils but instead seems to look for the name of the directory in `.stack/setup-cache` corresponding to the compiler.


Here is what I have 

```
max@max-VirtualBox:~/devel/sql-fragment$ !fin
find . -name cabal_macros.h
./.stack-work/dist/x86_64-linux/Cabal-1.18.1.5/build/autogen/cabal_macros.h
./.stack-work/dist/x86_64-linux/Cabal-1.22.5.0/build/autogen/cabal_macros.h
./dist/build/autogen/cabal_macros.h
./dist/dist-sandbox-16ceddde/build/autogen/cabal_macros.h

max@max-VirtualBox:~/devel/sql-fragment$ stack runghc get-cabal-configuration.hs  sql-fragment.cabal 
((build-directories "dist/build" "dist/build/autogen" "dist/build/generate-fragment/generate-fragment-tmp" ".stack-work/dist/x86_64-linux/Cabal-1.22.1.1/build" ".stack-work/dist/x86_64-linux/Cabal-1.22.1.1/build/autogen" ".stack-work/dist/x86_64-linux/Cabal-1.22.1.1/build/generate-fragment/generate-fragment-tmp") (source-directories "src" "executables" "tests") (extensions) (languages "Haskell2010") (dependencies "HList" "HUnit" "QuickCheck" "base" "containers" "dimensional" "doctest" "filepath" "hspec" "lens" "mtl" "pretty" "regex-pcre" "split" "tagged" "time") (other-options) (autogen-directories "dist/build/autogen" ".stack-work/dist/x86_64-linux/Cabal-1.22.1.1/build/autogen"))

max@max-VirtualBox:~/devel/sql-fragment$ find . -name cabal_macros.h
./.stack-work/dist/x86_64-linux/Cabal-1.18.1.5/build/autogen/cabal_macros.h
./.stack-work/dist/x86_64-linux/Cabal-1.22.5.0/build/autogen/cabal_macros.h
./dist/build/autogen/cabal_macros.h
./dist/dist-sandbox-16ceddde/build/autogen/cabal_macros.h

max@max-VirtualBox:~/devel/sql-fragment$ ls ~/.stack/setup-exe-cache/x86_64-linux
setup-Simple-Cabal-1.18.1.5-ghc-7.8.4   setup-Simple-Cabal-1.22.5.0-ghc-7.10.3
setup-Simple-Cabal-1.22.4.0-ghc-7.10.2  setup-Simple-Cabal-1.22.8.0-ghc-7.10.3

```

As you can see, get-cabal-configurations recommends Cabal-1.22.1, I only have a directory for  Cabal-1.18.5 and Cabal-1.22.4. (Cabal-1.22.4 is for ghc-7.10.2) and the cabal version associated to the compiler used by this project (7.8.4) is Cabal-1.18.5.

running

```
max@max-VirtualBox:~/devel/sql-fragment$ stack path --dist-dir
.stack-work/dist/x86_64-linux/Cabal-1.18.1.5
```

returns the correct version of Cabal.

Basically, the problem update when you update cabal. flycheck seems to pickup the updated version, whereas , stack use the version of cabal at the time you installed the compiler.
